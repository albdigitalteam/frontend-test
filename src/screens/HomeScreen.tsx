import React, { useRef, useState, useEffect } from 'react';
import {
  FlatList,
  NativeSyntheticEvent,
  NativeScrollEvent,
  Dimensions,
} from 'react-native';
import FAIcon from 'react-native-vector-icons/MaterialIcons';
import { useDispatch, useSelector } from 'react-redux';
import { useNavigation } from '@react-navigation/core';
import { NativeStackNavigationProp } from '@react-navigation/native-stack';
import styled, { useTheme } from 'styled-components/native';

import Header from '../components/Header';
import NewPost from '../components/NewPost';
import Post from '../components/Post';

import { SET_USERS, UserState } from '../store/slices/usersSlice';
import {
  GET_COMMENTS,
  GET_POSTS,
  Post as PostProps,
  PostsState,
} from '../store/slices/postsSlice';
import { RootState } from '../store';
import { RootStackParamList } from '../routes';

type HomeScreenNavigationProp = NativeStackNavigationProp<
  RootStackParamList,
  'Home'
>;

const POST_WIDTH = Dimensions.get('window').width * 0.85;

const HomeScreen: React.FC = () => {
  const flatListRef = useRef<FlatList | null>(null);
  const [floatingButtonVisibility, setFloatingButtonVisibility] =
    useState<boolean>(false);
  const { colors } = useTheme();
  const { navigate } = useNavigation<HomeScreenNavigationProp>();
  const dispatch = useDispatch();
  const { posts } = useSelector<RootState, PostsState>(state => state.post);
  const { userLogged } = useSelector<RootState, UserState>(state => state.user);

  const handleScrolling = (e: NativeSyntheticEvent<NativeScrollEvent>) => {
    const {
      nativeEvent: {
        contentOffset: { x },
      },
    } = e;
    if (!floatingButtonVisibility && x > 50) {
      setFloatingButtonVisibility(true);
    } else if (floatingButtonVisibility && x < 50) {
      setFloatingButtonVisibility(false);
    }
  };

  useEffect(() => {
    dispatch(SET_USERS());
    dispatch(GET_POSTS());
    dispatch(GET_COMMENTS());
  }, []);

  return (
    <StyledContainer>
      <Header title="POSTGRAM" showUserIcon />
      <NewPost userLogged={userLogged} />
      <StyledFlatList
        data={posts}
        ref={flatListRef}
        showsHorizontalScrollIndicator={false}
        horizontal
        onScroll={handleScrolling}
        scrollEventThrottle={16}
        decelerationRate={'fast'}
        snapToAlignment="start"
        keyboardDismissMode={'on-drag'}
        snapToOffsets={[...Array(posts.length)].map(
          (x, i) => i * (POST_WIDTH + 10) + (i - 1) * 10,
        )}
        keyExtractor={item => `${item.title}-${item.id}`}
        renderItem={({ item }: { item: PostProps }) => (
          <Post
            data={item}
            onPress={() => navigate('Details', { postId: item.id })}
          />
        )}
      />

      {floatingButtonVisibility && (
        <StyledFloatingButtonContainer
          onPress={() =>
            flatListRef.current?.scrollToOffset({ animated: true, offset: 0 })
          }>
          <FAIcon name="keyboard-arrow-left" size={25} color={colors.primary} />
        </StyledFloatingButtonContainer>
      )}
    </StyledContainer>
  );
};

const StyledContainer = styled.View`
  flex: 1;
  background-color: ${({ theme: { colors } }) => colors.primary};
`;

const StyledFlatList = styled(FlatList as new () => FlatList<PostProps>).attrs({
  contentContainerStyle: {
    paddingTop: 16,
    paddingHorizontal: 10,
    paddingBottom: 16
  },
})``;

const StyledFloatingButtonContainer = styled.TouchableOpacity`
  width: 40px;
  height: 40px;
  border-radius: 40px;
  background-color: ${({ theme: { colors } }) => colors.quaternary};
  position: absolute;
  bottom: 10px;
  right: 10px;
  align-items: center;
  justify-content: center;
`;

export default HomeScreen;
