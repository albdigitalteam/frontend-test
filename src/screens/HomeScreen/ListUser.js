import React, { memo } from 'react';
import { FlatList } from 'react-native';
import styled from 'styled-components/native';

import PropTypes from 'prop-types';

import Label from '@/components/Label';
import colors from '@/theme/colors';

const ListUser = ({ users, isLoading }) => {
  const imageUsers = {
    1: require('@/assets/images/1.png'),
    2: require('@/assets/images/2.png'),
    3: require('@/assets/images/3.png'),
    4: require('@/assets/images/4.png'),
    5: require('@/assets/images/5.png'),
    6: require('@/assets/images/6.png'),
    7: require('@/assets/images/7.png'),
    8: require('@/assets/images/8.png'),
    9: require('@/assets/images/9.png'),
    10: require('@/assets/images/10.png')
  };
  const renderItem = ({ item: { id, name } }) => {
    return (
      <>
        <StyledViewImage>
          <StyledImage source={imageUsers[id]} />
          <Label
            width={85}
            numberOfLines={1}
            fontSize={12}
            marginLeft={10}
            textAlign='center'
            lineHeight={24}
          >
            {name}
          </Label>
        </StyledViewImage>
      </>
    );
  };

  return (
    <StyledContentUsers>
      <FlatList
        data={users}
        horizontal
        keyExtractor={item => item.id}
        showsHorizontalScrollIndicator={false}
        renderItem={renderItem}
        isLoading={isLoading}
      />
    </StyledContentUsers>
  );
};

const StyledContentUsers = styled.View`
  width: 100%;
  height: 100px;
  align-items: center;
  justify-content: center;
  padding: 8px;
  background-color: ${colors.COLOR_WHITE};
`;

const StyledImage = styled.Image`
  width: 60px;
  height: 60px;
  border-radius: 50px;
  margin-left: 8px;
`;

const StyledViewImage = styled.View`
  flex: 1;
  align-items: center;
  justify-content: center;
`;

ListUser.defaultProps = {
  users: [],
  isLoading: false
};

ListUser.propTypes = {
  isLoading: PropTypes.bool,
  users: PropTypes.arrayOf(
    PropTypes.shape({
      id: PropTypes.number,
      name: PropTypes.string,
      username: PropTypes.string,
      email: PropTypes.string
    })
  )
};

export default memo(ListUser);
