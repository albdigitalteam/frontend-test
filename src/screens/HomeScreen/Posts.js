import React, { memo, useCallback } from 'react';
import { FlatList, RefreshControl } from 'react-native';

import PropTypes from 'prop-types';

import colors from '@/theme/colors';
import Card from '@/components/Card';
import { useNavigation } from '@react-navigation/native';

const Posts = ({ posts, isLoading, onRefresh }) => {
  const { navigate } = useNavigation();

  const handleNavigate = useCallback(
    id => {
      navigate('Comments', { id });
    },
    [navigate]
  );

  const renderItem = ({ item: { id, title, name, nameName } }) => {
    return (
      <Card
        id={id}
        onPress={() => {
          handleNavigate(id);
        }}
        title={title}
        name={name}
        userName={nameName}
      />
    );
  };
  const reflashControl = useCallback(
    () => (
      <RefreshControl
        progressViewOffset={32}
        tintColor={colors.PRIMARY}
        colors={[colors.PRIMARY]}
        refreshing={isLoading}
        onRefresh={onRefresh}
      />
    ),
    [isLoading, onRefresh]
  );

  return (
    <>
      <FlatList
        data={posts}
        keyExtractor={item => item.id}
        refreshControl={reflashControl()}
        renderItem={renderItem}
        refreshing={isLoading}
      />
    </>
  );
};

Posts.defaultProps = {
  posts: [],
  isLoading: false,
  onRefresh: () => {}
};

Posts.propTypes = {
  isLoading: PropTypes.bool,
  posts: PropTypes.arrayOf(
    PropTypes.shape({
      id: PropTypes.number,
      title: PropTypes.string,
      body: PropTypes.string
    })
  ),
  onRefresh: PropTypes.func
};

export default memo(Posts);
