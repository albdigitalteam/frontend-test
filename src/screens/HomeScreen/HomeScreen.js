import React, { useEffect, useCallback } from 'react';
import styled from 'styled-components/native';

import { useDispatch, useSelector } from 'react-redux';

import Label from '@/components/Label';
import colors from '@/theme/colors';

import { LOADING_USERS } from '@/store/slices/userSlice';
import { LOADING_POSTS } from '@/store/slices/postSlice';

import ListUser from './ListUser';
import Posts from './Posts';

const HomeScreen = () => {
  const dispatch = useDispatch();
  const { usersData, isLoading, errorMessage } = useSelector(
    ({ users }) => users
  );

  const { postsData } = useSelector(({ posts }) => posts);

  const loadingData = useCallback(() => {
    dispatch(LOADING_USERS());
    dispatch(LOADING_POSTS());
  }, [dispatch]);

  useEffect(() => {
    loadingData();
  }, [loadingData]);

  return (
    <StyledContainer>
      {errorMessage ? (
        <Label
          textAlign='center'
          fontWeight={400}
          fontSize={12}
          marginBottom={8}
          color={colors.DANGER}
        >
          {errorMessage}
        </Label>
      ) : (
        <>
          <ListUser users={usersData} isLoading={isLoading} />
          <Posts
            posts={postsData}
            isLoading={isLoading}
            onRefresh={loadingData}
          />
        </>
      )}
    </StyledContainer>
  );
};

const StyledContainer = styled.View`
  flex: 1;
`;

export default HomeScreen;
