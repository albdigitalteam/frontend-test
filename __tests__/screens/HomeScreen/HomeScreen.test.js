import 'react-native';
import React from 'react';
import {
  matchSnapshotWithTheme,
  renderWithTheme
} from '@/helpers/functions/testing';

import HomeScreen from '@/screens/HomeScreen';

describe('tests in HomeScreen Screen', () => {
  test('matches snapshot', () => {
    matchSnapshotWithTheme(<HomeScreen />);
  });

  test('renders correctly', () => {
    renderWithTheme(<HomeScreen />);
  });
});
