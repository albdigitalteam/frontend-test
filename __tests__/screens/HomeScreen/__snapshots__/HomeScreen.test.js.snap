// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`tests in HomeScreen Screen matches snapshot 1`] = `
<View
  style={
    Array [
      Object {
        "flexBasis": 0,
        "flexGrow": 1,
        "flexShrink": 1,
      },
    ]
  }
>
  <View
    style={
      Array [
        Object {
          "alignItems": "center",
          "backgroundColor": "#FFF",
          "height": 100,
          "justifyContent": "center",
          "paddingBottom": 8,
          "paddingLeft": 8,
          "paddingRight": 8,
          "paddingTop": 8,
          "width": "100%",
        },
      ]
    }
  >
    <RCTScrollView
      data={Array []}
      disableVirtualization={false}
      getItem={[Function]}
      getItemCount={[Function]}
      horizontal={true}
      initialNumToRender={10}
      isLoading={false}
      keyExtractor={[Function]}
      maxToRenderPerBatch={10}
      onContentSizeChange={[Function]}
      onEndReachedThreshold={2}
      onLayout={[Function]}
      onMomentumScrollEnd={[Function]}
      onScroll={[Function]}
      onScrollBeginDrag={[Function]}
      onScrollEndDrag={[Function]}
      removeClippedSubviews={false}
      renderItem={[Function]}
      scrollEventThrottle={50}
      showsHorizontalScrollIndicator={false}
      stickyHeaderIndices={Array []}
      updateCellsBatchingPeriod={50}
      viewabilityConfigCallbackPairs={Array []}
      windowSize={21}
    >
      <View />
    </RCTScrollView>
  </View>
  <RCTScrollView
    data={Array []}
    disableVirtualization={false}
    getItem={[Function]}
    getItemCount={[Function]}
    horizontal={false}
    initialNumToRender={10}
    keyExtractor={[Function]}
    maxToRenderPerBatch={10}
    onContentSizeChange={[Function]}
    onEndReachedThreshold={2}
    onLayout={[Function]}
    onMomentumScrollEnd={[Function]}
    onScroll={[Function]}
    onScrollBeginDrag={[Function]}
    onScrollEndDrag={[Function]}
    refreshControl={
      <RefreshControlMock
        colors={
          Array [
            "#25265E",
          ]
        }
        onRefresh={[Function]}
        progressViewOffset={32}
        refreshing={false}
        tintColor="#25265E"
      />
    }
    refreshing={false}
    removeClippedSubviews={false}
    renderItem={[Function]}
    scrollEventThrottle={50}
    stickyHeaderIndices={Array []}
    updateCellsBatchingPeriod={50}
    viewabilityConfigCallbackPairs={Array []}
    windowSize={21}
  >
    <RCTRefreshControl />
    <View />
  </RCTScrollView>
</View>
`;
